src/macaron/slsa_analyzer/package_registry/npm_registry.py

```diff
@@ -0,0 +1,265 @@
+# Copyright (c) 2023 - 2023, Oracle and/or its affiliates. All rights reserved.
+# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/.
+
+"""The module provides abstractions for the npm package registry."""
+
+import json
+import logging
+from typing import NamedTuple
+from urllib.parse import SplitResult, urlunsplit
+
+import requests
+
+from macaron.config.defaults import defaults
+from macaron.errors import ConfigurationError, InvalidHTTPResponseError, InvalidPURLError
+from macaron.slsa_analyzer.build_tool.base_build_tool import BaseBuildTool
+from macaron.slsa_analyzer.build_tool.npm import NPM
+from macaron.slsa_analyzer.build_tool.yarn import Yarn
+from macaron.slsa_analyzer.package_registry.package_registry import PackageRegistry
+from macaron.util import send_get_http_raw
+
+logger: logging.Logger = logging.getLogger(__name__)
+
+
+class NPMRegistry(PackageRegistry):
+    """This class implements the npm package registry.
+
+    There is no complete and up-to-date API documentation for the npm registry and the endpoints
+    are discovered by manual inspection of links on https://www.npmjs.com.
+    """
+
+    def __init__(
+        self,
+        hostname: str | None = None,
+        attestation_endpoint: str | None = None,
+        request_timeout: int | None = None,
+        enabled: bool = True,
+    ) -> None:
+        """
+        Initialize the npm Registry instance.
+
+        Parameters
+        ----------
+        hostname : str
+            The hostname of the npm registry.
+        attestation_endpoint : str | None
+            The attestation REST API.
+        request_timeout : int | None
+            The timeout (in seconds) for requests made to the package registry.
+        enabled: bool
+            Shows whether making REST API calls to npm registry is enabled.
+        """
+        self.hostname = hostname or ""
+        self.attestation_endpoint = attestation_endpoint or ""
+        self.request_timeout = request_timeout or 10
+        self.enabled = enabled
+        super().__init__("npm Registry")
+
+    def load_defaults(self) -> None:
+        """Load the .ini configuration for the current package registry.
+
+        Raises
+        ------
+        ConfigurationError
+            If there is a schema violation in the ``npm registry`` section.
+        """
+        section_name = "package_registry.npm"
+        if not defaults.has_section(section_name):
+            self.enabled = False
+            return
+        section = defaults[section_name]
+
+        if not section.getboolean("enabled", fallback=True):
+            self.enabled = False
+            logger.debug("npm registry is disabled in section [{section_name}] of the .ini configuration file.")
```

Comment:
You probably didn't mean to use f-string here.
```suggestion
            logger.debug(
                "npm registry is disabled in section [%s] of the .ini configuration file.",
                section_name,
            )
```